sympy/utilities/memoization.py

```diff
@@ -1,7 +1,16 @@
+from __future__ import annotations
 from functools import wraps
+from typing import Callable, List, TypeVar, Protocol, cast
 
+T_co = TypeVar("T_co", covariant=True)
+T = TypeVar("T")
 
-def recurrence_memo(initial):
+class RecurrenceMemoFunc(Protocol[T_co]):
+    def __call__(self, n: int) -> T_co: ...
+    def cache_length(self) -> int: ...
+    def fetch_item(self, x): ...
+
+def recurrence_memo(initial: List[T]) -> Callable[[Callable[[int, List[T]], T]], RecurrenceMemoFunc[T]]:
```

Comment:
Okay now that you have made annotations much longer this line is too long and should be wrapped. Same below.